File: app/src/main/java/com/narcyber/mvpbasics/db/data/dao/UserDao.java
package com.narcyber.mvpbasics.db.data.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.narcyber.mvpbasics.model.User;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface UserDao {


    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Completable insertOrUpdate(User user);

    @Query("SELECT * FROM users ")
    Observable<List<User>> getAllUsers();

    @Query("SELECT * FROM users WHERE email IS :email AND password IS :password")
    Single<User> getUserByEmailAndPassword(String email, String password);

    @Query("SELECT * FROM users WHERE userName IS :userName")
    Single<User> getUserByUsername(String userName);

    @Query("SELECT * FROM users WHERE email IS :email")
    Single<User> getUserByEmail(String email);

    @Delete
    Single<Integer> deleteUser(User user);

    @Update
    Completable update(User user);


}

File: app/src/main/java/com/narcyber/mvpbasics/db/data/dao/UserDao.kt
package com.narcyber.mvpbasics.db.data.dao

import androidx.room.*
import com.narcyber.mvpbasics.model.User
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single

@Dao
interface UserDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOrUpdate(user: User?): Completable

    @Query("SELECT * FROM users ")
    fun getAllUsers(): Observable<List<User>>

    @Query("SELECT * FROM users WHERE email IS :email AND password IS :password")
    fun getUserByEmailAndPassword(email: String, password: String): Single<User>

    @Query("SELECT * FROM users WHERE userName IS :userName")
    fun getUserByUsername(userName: String): Single<User>

    @Query("SELECT * FROM users WHERE email IS :email")
    fun getUserByEmail(email: String): Single<User>

    @Delete
    fun deleteUser(user: User): Single<Int>

    @Update
    fun update(user: User): Completable
}